docs/code-organisation-guidelines.md

```diff
@@ -0,0 +1,65 @@
+as a series of small easy-to-review commits that can be merged relatively quickly.
+
+* The `cardano-sl` git repository includes a number of distinct Haskell packages like `core`, `util`,
+  `explorer`, `networking` and so on. All Haskell modules within each of these packages should be
+  given a unique name prefix. All modules within `core` would have a prefix `Pos.Core`, modules
+  within `util` would have a name prefix `Pos.Util`, modules within `networking` might have a prefix
+  of `Pos.Network` and so on. This means that when looking at the imports for a module, it is
+  immediately obvious which package each import comes from. As an example of a file that needs to be
+  renamed/moved the file `core/Pos/Aeson/Core.hs` should be renamed to `core/Pos/Core/Aeson.hs` so
+  that if it is imported in another module its name `Pos.Core.Aeson` would immediately suggest its
+  location in the source tree.
+
+* All library code for a Haskell packages should be in the `Pos` directory of that package.
+  Currently some packages use `src` and then `src` has both `Pos` and `Test` sub-directories. Since
+  all test code should be in the `test` sub directory (see below), we should choose `Pos` over
+  `src/Pos`. For example `lib` currently has testing related code in the `src` directory and if
+  that test code moves to the `test` directory then there is no need for `src`.
+
+* Tests for a Haskell package should always be a part of the same package as the implementation of
+  the code it is testing. Test code should not be included in the library that a package provides.
+  All test code for a Haskell package should be in the `test` directory of that package and modules
+  with testing related code should have module names staring with `Test` and then reflecting the
+  Haskell package which the code is testing, so test code for the `core` package would have all its
+  test module names begin with `Test.Pos.Core`. An example of where the current code base
+  contravenes these guidelines is that `core` and `txp` currently don't define any tests, but at
+  least some of the functionality in these two modules are tested in `lib` package.
+
+* Arbitrary instances for testing should never be in the library source tree and should not be
+  compiled into the library defined by the package. Instead, Arbitrary instances should be under the
+  `test` directory as per the previous point. When Arbitrary instances or other testing helper
+  functionality needs to be used by other packages, a test specific package should be created. For
+  instance, the `core` package may have `Arbitrary` instances that are need for tests in the `txp`
+  package. In this case, there will be a cabal file `core/test/cardano-sl-core-test.cabal` which
+  exports the required modules, all of which would have a `Test.Pos.Core` name prefix.
+  [PR #2588](https://github.com/input-output-hk/cardano-sl/pull/2588) is an example of doing this
+  for the `crypto` package.
```

Comment:
I'm not convinced this is a hard and fast rule. I think we should default to providing non-orphan instances wherever possible -- on a codebase like this, it can be *very* hard to figure out where an instance is defined, and it's not like we're avoiding a `QuickCheck` transitive dependency (and, even if we were, QuickCheck doesn't bring in that many dependencies these days).

If we entirely shifted to a `hedgehog` (or other non type class approach), then I can see the appeal to locating the generators for a type in another package section or package.

Reply:
I share your distaste for orphan instances, but the current code base has `-fno-warn-orphans` in every cabal file and uses orphan instances even when I think it doesn't need to. IMO, one of the few legitimate uses of orphan instances is `Arbitrary` instances so you don't have to put them in the library itself.

I would be totally behind moving to Hedgehog. Amy last employer was Ambiata whose testing practices led to the development of Hedgehog.

Reply:
My understanding is that the argument for libraries is to avoid requiring that your downstream consumers incur a dependency that's only useful for testing, and so the orphan instance is fine (literally unobservable as you can't import test code).

However, that punts the orphan instance to all applications or downstream consumers of that library. Then, you get something like `quickcheck-instances` with a ton of dependencies that you can rope in that provides orphans, but in a canonical location. This is *also* not a problem, because presumably this is only imported into `test`s again.

When you're writing an application, you don't have to worry about "downstream," and you're going to incur the QuickCheck dependency regardless, so you might as well avoid the orphans. I don't know if there are any downsides to having arbitrary instances along the types in an application.

I think the cardano-sl codebase falls more into the "application" side of things than "library" at the moment, which explains my preference

Reply:
Purging the codebase of orphans would be a good bullet on the list, too, imo :)

Reply:
Uhm, that's a thorny one, and I can see valid arguments on both fronts. What I can bring to the table is my experience as a user of the Cardano libraries, both as a library writer and as an application writer (in the context of `wallet-new`):

- I share the distaste for orphans as well, with the notable exception that if they are provided as part of an application (as in an `executable` at the top of the "food chain") I tolerate them, otherwise I tend to avoid them.

- For the "classics" (`Eq`, `Ord`, `Show`, etc) the choice is obvious, but I have found the line getting more and more blurred when we tackle things like `Arbitrary` (that, these days, I consider at the same level of the "classics" and which I prefer to have non-orphaned) or even recently things like the Swagger schema (cfr `wallet-new`), which is, imho, a good example. When I started the work on `wallet-new` I have put in _bona fide_ all the swagger-related instances in a separate module (not even inside the `library`, but in the `executable`) because I wanted to separate the documentation concern with the rest. In this optic I totally see @erikd 's point of view.

- At first I didn't like too much the split in use in the code base at the moment (with modules like `Bi`, `Aeson` etc providing all orphans) but I can also see the other side of the coin: probably as the codebase grows and as more instances gets defined, I'm not sure up to which point putting everything in the same module where you define your types it's gonna scale (I genuinely don't know, that's it). My suspect is that soon one will lose sight of the "meat" of the module (the bare types) as that would be swallowed by the "white noise" generated by the instances.

Summarising, I think my current feeling is: I don't like orphans, but for large codebases they might become a necessity at some point? (you tell me 😁 ) Provided we have an unambiguous, well-defined and clash-free mechanism for providing that, I think I can live with that 😉

Reply:
At Ambiata (which IMHO had some of the best coding practices I've ever seen) we had a rule, that the only acceptable orphans were for `Arbitrary` instances (to keep them out of the library) and then we ended up writing our own property testing framework (which then became Hedgehog) that didn't use any typeclass mechanisms for generating data.

Reply:
I have heard good things about Ambiata's coding practice :)